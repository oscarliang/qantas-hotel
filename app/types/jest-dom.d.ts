// Extend Jest types with custom matchers
import '@testing-library/jest-dom'; 