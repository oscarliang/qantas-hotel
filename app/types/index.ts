export type HotelData = {
    results: Hotel[];
};

export type Hotel = {
    id: string;
    property: {
        propertyId: string;
        title: string;
        address: string[];
        previewImage: {
            url: string;
            caption: string;
            imageType: string;
        };
        rating: {
            ratingValue: number;
            ratingType: 'star' | 'self';
        };
    };
    offer: {
        promotion?: {
            title: string;
            type: string;
        };
        name: string;
        displayPrice: {
            amount: number;
            currency: string;
        };
        savings?: {
            amount: number;
            currency: string;
        } | null;
        cancellationOption: {
            cancellationType: 'FREE_CANCELLATION' | 'NOT_REFUNDABLE';
        };
    };
}; 