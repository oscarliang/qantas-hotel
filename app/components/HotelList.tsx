import React, { useState, useMemo } from 'react';
import HotelCard from './HotelCard';
import SortDropdown from './SortDropdown';
import { Hotel } from '../types';

type HotelListProps = {
  hotels: Hotel[];
};

type SortOption = 'high-low' | 'low-high';

const HotelList: React.FC<HotelListProps> = ({ hotels }) => {
  const [sortOption, setSortOption] = useState<SortOption>('high-low');

  const sortedHotels = useMemo(() => {
    if (!hotels) return [];

    return [...hotels].sort((a, b) => {
      const priceA = a.offer.displayPrice.amount;
      const priceB = b.offer.displayPrice.amount;

      return sortOption === 'high-low'
        ? priceB - priceA // high to low
        : priceA - priceB; // low to high
    });
  }, [hotels, sortOption]);

  if (!hotels || hotels.length === 0) {
    return <div className="py-8 text-center">No hotels found.</div>;
  }

  return (
    <div>
      <div className="mb-4 flex items-center justify-between">
        <div className="text-sm">
          <span className="font-bold">{hotels.length}</span>
          <span className="italic text-gray-600"> hotels in </span>
          <span className="font-bold">Sydney</span>.
        </div>
        <SortDropdown currentSort={sortOption} onSortChange={setSortOption} />
      </div>

      <div className="hotel-list">
        {sortedHotels.map((hotel) => (
          <HotelCard
            key={hotel.id}
            id={hotel.id}
            propertyId={hotel.property.propertyId}
            title={hotel.property.title}
            address={hotel.property.address}
            imageUrl={hotel.property.previewImage.url}
            imageCaption={hotel.property.previewImage.caption}
            rating={hotel.property.rating}
            promotion={hotel.offer.promotion}
            roomName={hotel.offer.name}
            price={hotel.offer.displayPrice}
            savings={hotel.offer.savings}
            cancellationType={hotel.offer.cancellationOption.cancellationType}
          />
        ))}
      </div>
    </div>
  );
};

export default HotelList;
