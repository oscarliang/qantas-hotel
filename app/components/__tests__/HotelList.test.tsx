import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import HotelList from '../HotelList';
import { Hotel } from '../../types';

// Mock the child components
jest.mock('../HotelCard', () => {
    return jest.fn(({ title }: { title: string }) => (
        <div data-testid="hotel-card">{title}</div>
    ));
});

jest.mock('../SortDropdown', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return jest.fn(({ currentSort, onSortChange }: { currentSort: string; onSortChange: (sort: any) => void }) => (
        <div data-testid="sort-dropdown">
            <select
                data-testid="sort-select"
                value={currentSort}
                onChange={(e) => onSortChange(e.target.value)}
            >
                <option value="high-low">Price high-low</option>
                <option value="low-high">Price low-high</option>
            </select>
        </div>
    ));
});

describe('HotelList Component', () => {
    const mockHotels: Hotel[] = [
        {
            id: 'hotel1',
            property: {
                propertyId: 'P1',
                title: 'Expensive Hotel',
                address: ['Test Address 1'],
                previewImage: {
                    url: '/test1.jpg',
                    caption: 'Test Caption 1',
                    imageType: 'PRIMARY'
                },
                rating: {
                    ratingValue: 5,
                    ratingType: 'star'
                }
            },
            offer: {
                name: 'Deluxe Room',
                displayPrice: {
                    amount: 500,
                    currency: 'AUD'
                },
                savings: null,
                cancellationOption: {
                    cancellationType: 'FREE_CANCELLATION'
                }
            }
        },
        {
            id: 'hotel2',
            property: {
                propertyId: 'P2',
                title: 'Cheap Hotel',
                address: ['Test Address 2'],
                previewImage: {
                    url: '/test2.jpg',
                    caption: 'Test Caption 2',
                    imageType: 'PRIMARY'
                },
                rating: {
                    ratingValue: 3,
                    ratingType: 'self'
                }
            },
            offer: {
                name: 'Standard Room',
                displayPrice: {
                    amount: 200,
                    currency: 'AUD'
                },
                savings: {
                    amount: 50,
                    currency: 'AUD'
                },
                cancellationOption: {
                    cancellationType: 'NOT_REFUNDABLE'
                }
            }
        }
    ];

    it('renders hotel count correctly', () => {
        const { container } = render(<HotelList hotels={mockHotels} />);

        expect(screen.getByText('2')).toBeInTheDocument(); // 2 hotels
        expect(screen.getByText('hotels in')).toBeInTheDocument();
        expect(screen.getByText('Sydney')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('renders sort dropdown with default high-low value', () => {
        render(<HotelList hotels={mockHotels} />);

        const sortSelect = screen.getByTestId('sort-select');
        expect(sortSelect).toHaveValue('high-low');
    });

    it('sorts hotels high to low by default', () => {
        render(<HotelList hotels={mockHotels} />);

        const hotelCards = screen.getAllByTestId('hotel-card');
        expect(hotelCards).toHaveLength(2);

        // First card should be the expensive hotel
        expect(hotelCards[0].textContent).toBe('Expensive Hotel');

        // Second card should be the cheap hotel
        expect(hotelCards[1].textContent).toBe('Cheap Hotel');
    });

    it('sorts hotels low to high when sort option changes', () => {
        render(<HotelList hotels={mockHotels} />);

        // Change sort to low-high
        const sortSelect = screen.getByTestId('sort-select');
        fireEvent.change(sortSelect, { target: { value: 'low-high' } });

        const hotelCards = screen.getAllByTestId('hotel-card');

        // Order should be reversed now
        expect(hotelCards[0].textContent).toBe('Cheap Hotel');
        expect(hotelCards[1].textContent).toBe('Expensive Hotel');
    });

    it('renders empty state when no hotels are provided', () => {
        const { container } = render(<HotelList hotels={[]} />);

        expect(screen.getByText('No hotels found.')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
}); 