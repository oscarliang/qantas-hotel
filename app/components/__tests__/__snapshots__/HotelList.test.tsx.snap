// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelList Component renders empty state when no hotels are provided 1`] = `
<div>
  <div
    class="py-8 text-center"
  >
    No hotels found.
  </div>
</div>
`;

exports[`HotelList Component renders hotel count correctly 1`] = `
<div>
  <div>
    <div
      class="mb-4 flex items-center justify-between"
    >
      <div
        class="text-sm"
      >
        <span
          class="font-bold"
        >
          2
        </span>
        <span
          class="italic text-gray-600"
        >
           hotels in 
        </span>
        <span
          class="font-bold"
        >
          Sydney
        </span>
        .
      </div>
      <div
        data-testid="sort-dropdown"
      >
        <select
          data-testid="sort-select"
        >
          <option
            value="high-low"
          >
            Price high-low
          </option>
          <option
            value="low-high"
          >
            Price low-high
          </option>
        </select>
      </div>
    </div>
    <div
      class="hotel-list"
    >
      <div
        data-testid="hotel-card"
      >
        Expensive Hotel
      </div>
      <div
        data-testid="hotel-card"
      >
        Cheap Hotel
      </div>
    </div>
  </div>
</div>
`;
